### PR TITLE
fix(bazel): Handle comment and string parsing

### LINKED ISSUE #8785: Parsing Bug in Bazel Extract
Hey awesome Renovate folks,

I was trying to track down a misparsing error that showed up in a Renovate PR. Along the way, I noticed what I'm pretty sure is a bug in the parser--though perhaps not the one causing my issue.

In particular, the comment says the parser should "if ["""] found then ignore all [)] until closing ["""] parsed." but doesn't seem to actually ignore ')' inside of a doctoring comment. Not sure how much this really matters, but it definitely doesn't do as the comment says.

A test case that shows the issue is `findBalancedParenIndex('""")""")')`. It returns -1, but should return 7 (the last index) after ignoring the first paren in the docstring. Could also add an assert checking that the parenNestingDepth is never negative.

Here's the function in question:
https://github.com/renovatebot/renovate/blob/d2a71472223ce2cae3ffaa5ade6ad5f8c1a29423/lib/manager/bazel/extract.ts#L56:58

I think this is different from the parsing problem causing my real issue. In the real issue, Renovate tries to update an external dependency in a comment (which is nice, actually), but misparses somewhere and tries to eat the next function definition. The code in extract.ts looks to me like it handles this fine--I'm not sure where in the code is causing this bug.

Easiest way to show the error case was a screenshot:

Screen Shot 2021-02-20 at 1 08 44 AM

Raw text for debugging:
```
    # Eigen already had a good BUILD file from Tensorflow.
# http_archive(
#     name = "rules_foreign_cc",
#     url = "https://github.com/bazelbuild/rules_foreign_cc/archive/dfccdce2c9d1063c59ddd331b94eb7cb528a96ee.tar.gz",
#     sha256 = "5469ef8b4e2c475de443c13290cf91ba7d1255899442b1e42fcb7fcdee8ceed8",
#     strip_prefix = "rules_foreign_cc-dfccdce2c9d1063c59ddd331b94eb7cb528a96ee",
# )
# load("@rules_foreign_cc//:workspace_definitions.bzl", "rules_foreign_cc_dependencies")
# rules_foreign_cc_dependencies()
# # Usage is a little weird, and depends 
# FOREIGN_CC_EXPOSE_ALL_FILES = """filegroup(name = "all", srcs = glob(["**"]), visibility = ["//visibility:public"])"""


########################################
# C++ & Cross-Platform Libraries

# Boost.
# Famous C++ library that gives rise to many new additions in the C++ standard library.
# See https://github.com/nelhage/rules_boost, recommended from https://docs.bazel.build/versions/master/rules.html
http_archive(
    name = "com_github_nelhage_rules_boost",
    url = "https://github.com/nelhage/rules_boost/archive/98495a618246683c9058dd87c2c78a2c06087999.tar.gz",
    sha256 = "f92cb7ed66a5b24f97a7fc3917407f808c70d2689273bdd68f93d70a379d22d3",
    strip_prefix = "rules_boost-98495a618246683c9058dd87c2c78a2c06087999",
)
load("@com_github_nelhage_rules_boost//:boost/boost.bzl", "boost_deps")
boost_deps() # Also pulls in a bunch of boost depenencies if you don't have them already. See https://github.com/nelhage/rules_boost/blob/master/boost/boost.bzl
```

Update: Yep, some different bug at work. findBalancedParenthases finds the right thing there.

Happy to help track down more tm, but it's late enough in my time zone that I better head to bed for now.

Thanks so much,
Chris

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

Use `moo` lexer to handle sophisticated parsing edge-cases (comments and strings).

## Context:

Closes #8785

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/bazel/extract.ts [hunk 1]
```diff
@@ -53,63 +54,96 @@ function parseUrl(urlString: string): UrlParsedResult | null {
...
+  };
 
-function parseContent(content: string): string[] {
-  return [
-    'container_pull',
-    'http_archive',
-    'http_file',
-    'go_repository',
-    'git_repository',
-  ].reduce(
```
Comment (zharinov) [inline lib/manager/bazel/extract.ts:98]: This iteration caused so huge diff on rewritten snapshot (note other snapshot tests are same).

**Comment by zharinov** [comment]
Can someone please point to good real use-case to test against?

**Review by viceice — DISMISSED** [review]
Looks good if it works ☺️

**Comment by zharinov** [comment]
Example repo: https://github.com/renovate-testing/cockroach-gen/pulls

Works similarly this branch and `master`. I guess, this one should work good enough to fix the original issue and not break existing behavior.

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
@zharinov I think this is the second time that this PR has failed tests with out of memory limits. Is there a chance it's caused by the code in this branch?

**Comment by zharinov** [comment]
I've noticed submodules test timeout for another branch. But for memory allocations, I'm not sure.

**Comment by viceice** [comment]
I've seen this on my pr too. Maybe we should enable jest mem stats again.

**Comment by zharinov** [comment]
image

**Comment by zharinov** [comment]
This PR is cursed, I guess

**Pull request closed without merging**